torch/testing/_internal/common_nn.py

```diff
@@ -4886,6 +4901,37 @@ def __init__(self, constructor, desc='', reference_fn=None, fullname=None, **kwa
         self._extra_kwargs = kwargs
         self._arg_cache = {}
 
+        if self.is_criterion_test:
```

Comment:
This part makes me worry about trying to merge all these classes upfront.

The OpInfo class explicitly lists all the allowed metadata kwargs and each is documented, but people still complain we don't have enough documentation! In this case parsing args out of the kwargs dict might be hard to understand. 

More generally, trying to merge these classes is going to mean inheriting both the current classes, their instantiations, and their tests. That might be hard to change at once? Would it be simpler to start with a new ModuleInfo class, one instantiation of it, and one test the replaces that existing tests? Then, instead of merging the current tests into one thing, the old class types could be ported to the new ModuleInfo type, until finally the old classes and tests could just be deleted.

Reply:
> More generally, trying to merge these classes is going to mean inheriting both the current classes, their instantiations, and their tests. That might be hard to change at once?

Indeed it was tricky, but I did it carefully & incrementally :) The purpose of the PR is to validate that this was done correctly in a way that maintains existing test coverage.

> Would it be simpler to start with a new ModuleInfo class, one instantiation of it, and one test the replaces that existing tests?

I'm also doing this, currently in a separate branch. I did not find it simpler to start with that without some refactoring on the existing code side. The refactoring helps identify what is actually being tested to inform the new tests & the new `ModuleInfo` abstraction + module tests will be easier to review when they can be compared against a more compact old test structure.